File: record_android/android/src/main/java/com/llfbandit/record/record/AudioRecorder.java
package com.llfbandit.record.record;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.llfbandit.record.record.processor.AacProcessor;
import com.llfbandit.record.record.processor.AmrProcessor;
import com.llfbandit.record.record.processor.AudioProcessor;
import com.llfbandit.record.record.processor.FlacProcessor;
import com.llfbandit.record.record.processor.OggProcessor;
import com.llfbandit.record.record.processor.OpusProcessor;
import com.llfbandit.record.record.processor.PcmProcessor;
import com.llfbandit.record.record.processor.WavProcessor;
import com.llfbandit.record.record.stream.RecorderRecordStreamHandler;
import com.llfbandit.record.record.stream.RecorderStateStreamHandler;

import java.util.ArrayList;
import java.util.List;

public class AudioRecorder implements AudioProcessor.OnAudioProcessorListener {
  private static final String TAG = "AudioRecorder";

  // Recording config
  private final RecordConfig config;
  // Recorder streams
  private final RecorderStateStreamHandler recorderStateStreamHandler;
  private final RecorderRecordStreamHandler recorderRecordStreamHandler;

  @Nullable
  private AudioProcessor audioProcessor;

  private double amplitude = -160.0;
  // Max amplitude recorded
  private double maxAmplitude = -160.0;

  public AudioRecorder(
      @NonNull RecordConfig config,
      @NonNull RecorderStateStreamHandler recorderStateStreamHandler,
      @NonNull RecorderRecordStreamHandler recorderRecordStreamHandler
  ) {
    this.config = config;
    this.recorderStateStreamHandler = recorderStateStreamHandler;
    this.recorderRecordStreamHandler = recorderRecordStreamHandler;
  }

  public void dispose() {
    stop();
  }

  /**
   * Starts the recording with the given config.
   */
  public void start() throws Exception {
    audioProcessor = selectProcessor();

    audioProcessor.start(config);
  }

  /**
   * Stops the recording.
   */
  public void stop() {
    if (audioProcessor != null) {
      audioProcessor.stop();
    }
  }

  /**
   * Pauses the recording if currently running.
   */
  public void pause() {
    if (audioProcessor != null) {
      audioProcessor.pause();
    }
  }

  /**
   * Resumes the recording if currently paused.
   */
  public void resume() {
    if (audioProcessor != null) {
      audioProcessor.resume();
    }
  }

  /**
   * Gets the state the of recording
   *
   * @return True if recording. False otherwise.
   */
  public boolean isRecording() {
    if (audioProcessor != null) {
      return audioProcessor.isRecording();
    }

    return false;
  }

  /**
   * Gets the state the of recording
   *
   * @return True if paused. False otherwise.
   */
  public boolean isPaused() {
    if (audioProcessor != null) {
      return audioProcessor.isPaused();
    }

    return false;
  }

  /**
   * Gets the amplitude
   *
   * @return List with current and max amplitude values
   */
  public List<Double> getAmplitude() {
    List<Double> amps = new ArrayList<>();

    amps.add(amplitude);
    amps.add(maxAmplitude);

    return amps;
  }

  @Override
  public void onRecord() {
    updateState(RecordState.RECORD);
  }

  @Override
  public void onPause() {
    updateState(RecordState.PAUSE);
  }

  @Override
  public void onStop() {
    updateState(RecordState.STOP);
  }

  @Override
  public void onFailure(Exception ex) {
    Log.e(TAG, ex.getMessage(), ex);

    recorderStateStreamHandler.sendStateErrorEvent(ex);

    try {
      stop();
    } catch (Exception ignored) {
    }
  }

  @Override
  public void onAmplitude(int amplitude) {
    this.amplitude = amplitude;

    if (amplitude > maxAmplitude) {
      maxAmplitude = amplitude;
    }
  }

  @Override
  public void onAudioChunk(byte[] chunk) {
    recorderRecordStreamHandler.sendRecordChunkEvent(chunk);
  }

  private AudioProcessor selectProcessor() throws Exception {
    switch (config.encoder) {
      case "aacLc":
      case "aacEld":
      case "aacHe":
        return new AacProcessor(this);
      case "amrNb":
      case "amrWb":
        return new AmrProcessor(this);
      case "flac":
        return new FlacProcessor(this);
      case "pcm16bit":
      case "pcm8bit":
        return new PcmProcessor(this);
      case "opus":
        return new OpusProcessor(this);
      case "vorbisOgg":
        return new OggProcessor(this);
      case "wav":
        return new WavProcessor(this);
    }

    throw new Exception("Unknown encoder: " + config.encoder);
  }

  private void updateState(RecordState state) {
    recorderStateStreamHandler.sendStateEvent(state.id);
  }
}

File: record_android/android/src/main/kotlin/com/llfbandit/record/record/AudioRecorder.kt
package com.llfbandit.record.record

import android.util.Log
import com.llfbandit.record.record.stream.RecorderRecordStreamHandler
import com.llfbandit.record.record.stream.RecorderStateStreamHandler

interface OnAudioRecordListener {
    fun onRecord()
    fun onPause()
    fun onStop()
    fun onFailure(ex: Exception)
    fun onAudioChunk(chunk: ByteArray)
}

class AudioRecorder(
    // Recorder streams
    private val recorderStateStreamHandler: RecorderStateStreamHandler,
    private val recorderRecordStreamHandler: RecorderRecordStreamHandler
) : OnAudioRecordListener {
    companion object {
        private val TAG = AudioRecorder::class.java.simpleName
    }

    // Recorder thread with which we will interact
    private var recorderThread: RecordThread? = null
    // Amplitude
    private var maxAmplitude = -160.0
    // Recording config
    private var config: RecordConfig? = null
    // Stop callback to be synchronized between stop method return & record stop
    private var stopCb: ((path: String?) -> Unit)? = null

    /**
     * Starts the recording with the given config.
     */
    @Throws(Exception::class)
    fun start(config: RecordConfig) {
        this.config = config

        recorderThread = RecordThread(config, this)
        recorderThread!!.start()
    }

    /**
     * Stops the recording.
     */
    fun stop(stopCb: ((path: String?) -> Unit)?) {
        this.stopCb = stopCb

        recorderThread?.stopRecording()
    }

    /**
     * Stops the recording and delete file.
     */
    fun cancel() {
        recorderThread?.cancelRecording()
    }

    /**
     * Pauses the recording if currently running.
     */
    fun pause() {
        recorderThread?.pauseRecording()
    }

    /**
     * Resumes the recording if currently paused.
     */
    fun resume() {
        recorderThread?.resumeRecording()
    }

    /**
     * Gets the state the of recording
     *
     * @return True if recording. False otherwise.
     */
    val isRecording: Boolean
        get() = recorderThread?.isRecording() == true

    /**
     * Gets the state the of recording
     *
     * @return True if paused. False otherwise.
     */
    val isPaused: Boolean
        get() = recorderThread?.isPaused() == true

    /**
     * Gets the amplitude
     *
     * @return List with current and max amplitude values
     */
    fun getAmplitude(): List<Double> {
        val amplitude = recorderThread?.getAmplitude() ?: -160.0
        val amps: MutableList<Double> = ArrayList()
        amps.add(amplitude)
        amps.add(maxAmplitude)
        return amps
    }

    fun dispose() {
        stop(null)
    }

    // OnAudioRecordListener
    override fun onRecord() {
        recorderStateStreamHandler.sendStateEvent(RecordState.RECORD.id)
    }

    override fun onPause() {
        recorderStateStreamHandler.sendStateEvent(RecordState.PAUSE.id)
    }

    override fun onStop() {
        stopCb?.invoke(config?.path)
        stopCb = null

        recorderStateStreamHandler.sendStateEvent(RecordState.STOP.id)
    }

    override fun onFailure(ex: Exception) {
        Log.e(TAG, ex.message, ex)
        recorderStateStreamHandler.sendStateErrorEvent(ex)
    }

    override fun onAudioChunk(chunk: ByteArray) {
        recorderRecordStreamHandler.sendRecordChunkEvent(chunk)
    }
}